### PR TITLE
[TextControls] Add accessibilityLabel override to MDCBaseTextField

### LINKED ISSUE #8771: [TextControls] Internal issue: b/144275532
This was filed as an internal issue. If you are a Googler, please visit [b/144275532](http://b/144275532) for more details.

---
#### Internal data
- Associated internal bug: [b/144275532](http://b/144275532)

### PR DESCRIPTION
**Pull request by andrewoverton**
UITextField has some VoiceOver support built in--the text the user enters is reflected in the accessibilityValue. The accessibilityLabel, however, seems to be nil unless a user specifies something for it. This PR doesn't change that behavior, but it does make it so that if no accessibilityLabel is set the textfield returns a concatenated string consisting of the label text and any text from the assistive labels.

Closes #8771.

### CODE REVIEW

File: components/TextControls/src/MDCBaseTextField.m [hunk 1]
```diff
@@ -662,6 +662,27 @@ - (MDCTextControlColorViewModel *)textControlColorViewModelForState:
   return colorViewModel;
 }
 
+#pragma mark Accessibility Overrides
+
+- (NSString *)accessibilityLabel {
```
Comment (jverkoey) [inline components/TextControls/src/MDCBaseTextField.m:667]: Please document this behavior in the header of this file.

File: components/TextControls/tests/unit/MDCBaseTextFieldTests.m [hunk 2]
```diff
@@ -338,4 +338,38 @@ - (void)testPlaceholderVisibility {
...
+  XCTAssertTrue([concatenatedText isEqualToString:textField.accessibilityLabel]);
+}
+
+- (void)testClientSpecifiedAccessibilityLabel {
+  // Given
+  CGRect textFieldFrame = CGRectMake(0, 0, 130, 100);
+  MDCBaseTextField *textField = [[MDCBaseTextField alloc] initWithFrame:textFieldFrame];
+  NSString *labelText = @"label text";
+  NSString *leadingAssistiveLabelText = @"leading assistive label text";
+  NSString *userSpecifiedAccessibilityLabel = @"user specified accessibility label";
```
Comment (jverkoey) [inline components/TextControls/tests/unit/MDCBaseTextFieldTests.m]: nit: user-specified

File: components/TextControls/tests/unit/MDCBaseTextFieldTests.m [hunk 3]
```diff
@@ -338,4 +338,38 @@ - (void)testPlaceholderVisibility {
...
+
+  // When
+  textField.label.text = labelText;
+  textField.leadingAssistiveLabel.text = leadingAssistiveLabelText;
+
+  // Then
+  NSString *concatenatedText =
+      [NSString stringWithFormat:@"%@, %@", labelText, leadingAssistiveLabelText];
+  XCTAssertTrue([concatenatedText isEqualToString:textField.accessibilityLabel]);
+}
```
Comment (jverkoey) [inline components/TextControls/tests/unit/MDCBaseTextFieldTests.m:426]: Please add tests with full coverage of the following:

- One test for each property where only one of the three properties is set.
- All three of the properties are set.
- None of the properties are set.
- At least one test where combinations of two properties are set.

File: components/TextControls/src/MDCBaseTextField.m (hunk 1 above)
Comment (andrewoverton) [reply components/TextControls/src/MDCBaseTextField.m:667]: Done.

File: components/TextControls/tests/unit/MDCBaseTextFieldTests.m (hunk 2 above)
Comment (andrewoverton) [reply components/TextControls/tests/unit/MDCBaseTextFieldTests.m]: Done.

File: components/TextControls/tests/unit/MDCBaseTextFieldTests.m (hunk 3 above)
Comment (andrewoverton) [reply components/TextControls/tests/unit/MDCBaseTextFieldTests.m:426]: Done.

File: components/TextControls/tests/unit/MDCBaseTextFieldTests.m [hunk 4]
```diff
@@ -338,4 +338,108 @@ - (void)testPlaceholderVisibility {
...
 }
 
+- (void)testDefaultAccessibilityLabelWithOnlyLabelText {
+  // Given
+  CGRect textFieldFrame = CGRectMake(0, 0, 130, 100);
+  MDCBaseTextField *textField = [[MDCBaseTextField alloc] initWithFrame:textFieldFrame];
+  NSString *labelText = @"label text";
+
+  // When
+  textField.label.text = labelText;
```
Comment (randallli) [inline components/TextControls/tests/unit/MDCBaseTextFieldTests.m:348]: Yes and let us test the attributedText too!

https://developer.apple.com/documentation/uikit/uitextfield/1619602-attributedtext?language=objc

**Review by randallli — APPROVED** [review]

File: components/TextControls/tests/unit/MDCBaseTextFieldTests.m (hunk 4 above)
Comment (randallli) [inline components/TextControls/tests/unit/MDCBaseTextFieldTests.m:348]: I was expecting textfield's text property to be used.
Why are we using to the label's text?

https://developer.apple.com/documentation/uikit/uitextfield/1619635-text?language=objc

**Review by randallli — APPROVED** [review]

File: components/TextControls/tests/unit/MDCBaseTextFieldTests.m (hunk 4 above)
Comment (andrewoverton) [reply components/TextControls/tests/unit/MDCBaseTextFieldTests.m:348]: Hmmm I don't want people to use the attributedText property on the label. Remember how I mentioned that I was considering making the label private and only exposing a pass through property `labelText`? This is why :/

Comment (andrewoverton) [reply components/TextControls/tests/unit/MDCBaseTextFieldTests.m:348]: The textfield's text property is accounted for by `accessibilityValue` and we're testing `accessibilityLabel` here. I guess we're going to have `accessibilityLabel` be a combination of the label text and the assistive label text and we're going to use the default accessibilityValue behavior.

**Pull request merged**